modules/osdk-monitoring-custom-metrics.adoc

```diff
@@ -0,0 +1,159 @@
++
+.`controllers/memcached_controller_test_metrics.go` file
+[%collapsible]
+====
+[source,go]
+----
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+)
+
+
+var (
+    widgets = prometheus.NewCounter(
+        prometheus.CounterOpts{
+            Name: "widgets_total",
+            Help: "Number of widgets processed",
+        },
+    )
+    widgetFailures = prometheus.NewCounter(
+        prometheus.CounterOpts{
+            Name: "widget_failures_total",
+            Help: "Number of failed widgets",
+        },
+    )
+)
+
+func init() {
+    // Register custom metrics with the global prometheus registry
+    metrics.Registry.MustRegister(widgets, widgetFailures)
+}
+----
+====
+
+. Record to these collectors from any part of the reconcile loop in the `main` controller class, which determines the business logic for the metric:
++
+.`controllers/memcached_controller.go` file
+[%collapsible]
+====
+[source,go]
+----
+func (r *MemcachedReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	...
+	...
+	// Add metrics
+	widgets.Inc()
+	widgetFailures.Inc()
+
+	return ctrl.Result{}, nil
+}
+----
+====
+
+. Deploy the Operator:
++
+[source,terminal]
+----
+$ make deploy
+----
+
+. Create role and role binding definitions to allow the service monitor of the Operator to be scraped by the Prometheus instance of the {product-title} cluster.
++
+Roles must be assigned so that service accounts have the permissions to scrape the metrics of the namespace:
++
+.`config/prometheus/role.yaml` role
+[%collapsible]
+====
+[source,yaml]
+----
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: prometheus-k8s-role
+  namespace: <operator_namespace>
+rules:
+  - apiGroups:
+      - ""
+    resources:
+      - endpoints
+      - pods
+      - services:q!
+      - nodes
+      - secrets
+    verbs:
+      - get
+      - list
+      - watch
+----
+====
++
+.`config/prometheus/rolebinding.yaml` role binding
+[%collapsible]
+====
+[source,yaml]
+----
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: prometheus-k8s-rolebinding
+  namespace: memcached-operator-system
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: prometheus-k8s-role
+subjects:
+  - kind: ServiceAccount
+    name: prometheus-k8s
+    namespace: openshift-monitoring
+----
+====
+
+. Apply the roles and role bindings for the deployed Operator:
++
+[source,terminal]
++
+----
+$ oc apply -f config/prometheus/role.yaml
+----
++
+[source,terminal]
+----
+$ oc apply -f config/prometheus/rolebinding.yaml
+----
+
+
+. Set the labels for the namespace that you want to scrape, which enables OpenShift cluster monitoring for that namespace:
```

Comment:
Note for reviewers: "OpenShift cluster monitoring" appears to be the commonly used phrase elsewhere in the repo, hence not going with `{product-title}` here for "OpenShift".